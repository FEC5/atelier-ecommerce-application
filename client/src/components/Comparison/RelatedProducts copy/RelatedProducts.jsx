import React from 'react';
import PubSub from 'pubsub-js';
// eslint-disable-next-line import/extensions
import Card from '../Card/Card.jsx';
import relatedProducts from './RelatedProducts.module.css';
import instance from '../../../../../axiosConfig.js';

// eslint-disable-next-line react/prefer-stateless-function
class RelatedProducts extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isRightEdgeInBound: true,
      leftOffset: 0,
      relatedProductIds: [],
    };
    this.handleScroll = this.handleScroll.bind(this);
    this.rightScrollerDisplay = this.rightScrollerDisplay.bind(this);
  }

  componentDidMount() {
    this.token = PubSub.subscribe('newProductId', (msg, data) => {
      this.updateRelatedProducts(data.id)
        .then(() => {
          this.rightScrollerDisplay();
        });
    });

    this.updateRelatedProducts(this.props.productId)
      .then(() => {
        this.rightScrollerDisplay();
      });
  }

  componentWillUnmount() {
    PubSub.unsubscribe(this.token);
  }

  handleScroll(isRight) {
    return async () => {
      const unitOffset = isRight ? -192 : 192;
      const { leftOffset } = this.state;
      await new Promise((resolve) => {
        this.setState({
          leftOffset: leftOffset + unitOffset,
        }, resolve);
      });
      this.rightScrollerDisplay();
    };
  }

  // if the right edge of the carousel is inside the container
  // set this.state.isRightEdgeInBound to false
  rightScrollerDisplay() {
    const { clientWidth: conainerWidth } = this.main || {};
    const { clientWidth: wrapperWidth } = this.carouselWrapper;
    const { leftOffset } = this.state;
    if (wrapperWidth + leftOffset <= conainerWidth) {
      this.setState({
        isRightEdgeInBound: true,
      });
    } else {
      this.setState({
        isRightEdgeInBound: false,
      });
    }
  }

  async updateRelatedProducts(productId) {
    try {
      const response = await instance.get(`/products/${productId}/related`);
      let { data } = response;
      data = [...new Set(data)];

      await new Promise((res) => {
        this.setState({
          relatedProductIds: data,
        }, res);
      });
    } catch (e) {
      console.warn(e);
    }
  }

  render() {
    const { leftOffset, isRightEdgeInBound, relatedProductIds } = this.state;
    const { productId, updateProductId } = this.props;
    return (
      <div className={relatedProducts['related-products']}>
        <div className={relatedProducts.title} data-testid="title">
          RELATED PRODUCTS
        </div>
        <span>
          Current Product Id:
          {productId}
        </span>
        <div className={relatedProducts.main} ref={(ele) => { this.main = ele; }}>
          <div
            className={relatedProducts['scroll-left']}
            style={{ display: leftOffset === 0 ? 'none' : 'flex' }}
          >
            <div
              className={relatedProducts['arrow-left']}
            />
            <button type="button" onClick={this.handleScroll(false)} />
          </div>
          <div
            className={relatedProducts['scroll-right']}
            style={{ display: isRightEdgeInBound ? 'none' : 'flex' }}
          >
            <div
              className={relatedProducts['arrow-right']}
            />
            <button type="button" onClick={this.handleScroll(true)} />
          </div>
          <div className={relatedProducts.carousel}>
            <div
              className={relatedProducts['carousel-wrapper']}
              ref={(ele) => { this.carouselWrapper = ele; }}
              style={{ left: leftOffset }}
            >
              {relatedProductIds.map((id) =>
                <Card key={id} id={id} updateProductId={updateProductId} />
              )}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default RelatedProducts;
