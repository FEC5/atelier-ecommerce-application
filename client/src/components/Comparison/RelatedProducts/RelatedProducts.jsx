import React from 'react';
// eslint-disable-next-line import/extensions
import Card from '../Card/Card.jsx';
import relatedProducts from './RelatedProducts.module.css';

// eslint-disable-next-line react/prefer-stateless-function
class RelatedProducts extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      leftOffset: 0,
      products: [1, 2, 3, 4, 5, 6, 6, 7],
    };
    this.handleScroll = RelatedProducts.handleScroll.bind(this);
  }

  static handleScroll(isRight) {
    return (e) => {
      const { leftOffset } = this.state;
      const cardWidth = this.refs.card;
      console.log(cardWidth);
      const unitOffset = isRight ? cardWidth : -cardWidth;
      this.setState({
        leftOffset: leftOffset + unitOffset,
      });
    };
  }

  render() {
    const { currCardIndex, products } = this.state;
    return (
      <div className="related-products">
        <div className={relatedProducts.title}>
          RELATED PRODUCTS
        </div>
        <div className={relatedProducts.main}>
          <div className={relatedProducts.left}>
            <div
              className={relatedProducts['arrow-left']}
              style={{ display: currCardIndex === 0 ? 'none' : 'inline-block' }}
            />
            <button type="button" onClick={this.handleScroll(false)} />
          </div>
          <div className={relatedProducts.right}>
            <div
              className={relatedProducts['arrow-right']}
              style={{ display: currCardIndex === products.length ? 'none' : 'inline-block' }}
            />
            <button type="button" onClick={this.handleScroll(true)} />
          </div>
          <div className={relatedProducts.carousel}>
            <div className={relatedProducts['carousel-wrapper']}>
              <Card ref='card'/>
              <Card />
              <Card />
              <Card />
              <Card />
              <Card />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default RelatedProducts;
