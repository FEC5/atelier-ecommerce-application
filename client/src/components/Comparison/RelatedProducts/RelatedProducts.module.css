.related-products {
  border: 3px solid black;
  width: 70%;
  height: 300px;
}

.main{
  display: flex;
  align-items: center;
  position: relative;
}

.carousel{
  z-index: 1;
  overflow: hidden;
  width:100%;
  display: flex;
  align-items: center;
}

.carousel-wrapper{
  white-space: nowrap;
  position: relative;
  width: auto;
  display: flex;
  align-items: center;
}


.scroll-left{
  z-index: 2;
  position: absolute;
  left: 20px;
  width:30px;
  height: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.scroll-right{
  z-index: 3;
  position: absolute;
  right: 20px;
  width:30px;
  height: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
}

button{
  position: absolute;
  height: 60px;
  width: 50px;
  opacity: 0.3;
  border-radius: 10%;
}

.arrow-left{
  box-sizing: border-box;
  width: 20px;
  height: 20px;
  border: solid lightgray;
  border-width: 0 6px 6px 0;
  transform: rotate(135deg);
  display: inline-block;
  position: absolute;
  left:8px
}

.arrow-right{
  box-sizing: border-box;
  width: 20px;
  height: 20px;
  border: solid lightgray;
  border-width: 0 6px 6px 0;
  transform: rotate(-45deg);
  display: inline-block;
  position: absolute;
  right: 8px
}

