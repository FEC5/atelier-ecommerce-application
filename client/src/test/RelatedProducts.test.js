/*
test example of a component
in command line run 'npm test'
*/
import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import RelatedProducts from '../components/Comparison/RelatedProducts/RelatedProducts.jsx';

afterEach(cleanup);

test('should take a snapshot', () => {
  const { asFragment } = render(<RelatedProducts />);

  expect(asFragment(<RelatedProducts />)).toMatchSnapshot();
});

// test if the dom element with test id 'title' in <RelatedProducts />
// have text content 'RELATED PRODUCTS'
test('should equal to RELATED PRODUCTS', () => {
  const { getByTestId } = render(<RelatedProducts />);
  expect(getByTestId('title')).toHaveTextContent('RELATED PRODUCTS');
});
