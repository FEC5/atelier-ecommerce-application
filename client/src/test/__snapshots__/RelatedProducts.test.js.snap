// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="related-products"
  >
    <div
      class="title"
      data-testid="title"
    >
      RELATED PRODUCTS
    </div>
    <div
      class="main"
    >
      <div
        class="scroll-left"
        style="display: none;"
      >
        <div
          class="arrow-left"
        />
        <button
          type="button"
        />
      </div>
      <div
        class="scroll-right"
        style="display: none;"
      >
        <div
          class="arrow-right"
        />
        <button
          type="button"
        />
      </div>
      <div
        class="carousel"
      >
        <div
          class="carousel-wrapper"
          style="left: 0px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
